// @mysten/seal/dist/esm/utils.js
import { fromHex, toHex } from "@mysten/bcs";
import { isValidSuiObjectId } from "@mysten/sui/utils";
import { UserError } from "./error.js";
function xor(a, b) {
  if (a.length !== b.length) {
    throw new Error("Invalid input");
  }
  return xorUnchecked(a, b);
}
function xorUnchecked(a, b) {
  return a.map((ai, i) => ai ^ b[i]);
}
function createFullId(dst, packageId, innerId) {
  if (!isValidSuiObjectId(packageId)) {
    throw new UserError(`Invalid package ID ${packageId}`);
  }
  const packageIdBytes = fromHex(packageId);
  const innerIdBytes = fromHex(innerId);
  const fullId = new Uint8Array(1 + dst.length + packageIdBytes.length + innerIdBytes.length);
  fullId.set([dst.length], 0);
  fullId.set(dst, 1);
  fullId.set(packageIdBytes, 1 + dst.length);
  fullId.set(innerIdBytes, 1 + dst.length + packageIdBytes.length);
  return toHex(fullId);
}
class Version {
  constructor(version) {
    const parts = version.split(".").map(Number);
    if (parts.length !== 3 || parts.some((part) => isNaN(part) || part < 0)) {
      throw new UserError(`Invalid version format: ${version}`);
    }
    this.major = parts[0];
    this.minor = parts[1];
    this.patch = parts[2];
  }
  // Compare this version with another version. True if this version is older than the other version.
  older_than(other) {
    if (this.major !== other.major) {
      return this.major < other.major;
    } else if (this.minor !== other.minor) {
      return this.minor < other.minor;
    }
    return this.patch < other.patch;
  }
}
export {
  Version,
  createFullId,
  xor,
  xorUnchecked
};
